torch/_inductor/codegen/cpp.py

```diff
@@ -2021,6 +2024,33 @@ def gen_loop(loop: LoopLevel):
 
             stack.enter_context(code.indent())
             if loop_nest.root:
+                if (
+                    has_outer_loop_kernel
+                    and isinstance(V.local_buffer_context, LocalBufferContext)
+                    and V.local_buffer_context.local_buffers
+                ):
+                    # Allocate local buffer
+                    local_buffers = V.local_buffer_context.local_buffers
+                    assert len(local_buffers.items()) == 1
+                    local_buffer = next(iter(local_buffers.items()))[1]
+                    # For dynamic size, rename s to ks
+                    local_buf_size = sympy_product(
+                        [
+                            self.rename_indexing(size_val)
+                            for size_val in local_buffer.get_layout().size
+                        ]
+                    )
+                    local_buf_dtype = DTYPE_TO_CPP[local_buffer.get_layout().dtype]
+                    allocate = (
+                        f"std::make_unique<{local_buf_dtype} []>({local_buf_size})"
```

Comment:
IIUC the local buffer won't actually be any smaller than the global buffer, it's just that you'll only use a smaller proportion of it in order to improve cache locality. Is that right?

Reply:
I think we actually change the size of buffer to be used. Task example of a softmax along the last dim:
  - Input size: [4, 12, 1025, 1024]
  - Running thread: 56

Calculate the buffer size used:

- Previously, we will allocate a global buffer with size `[4, 12, 1025, 1024]` and each thread will write/read a proportion of this global buffer.

- With this commit, we will allocate 56 (thread number) local buffers and each with size [1024] binding to one thread, and each thread will write/read its local buffer.